### DISCUSSION TITLE
Dio Interceptor

### DISCUSSION THREAD

**Discussion by scduarte** [Flutterando/forum#46] [Perguntas e Respostas]
Alguem sabe como faço para chamar novamente o método onRequest do interceptor, na versao antiga do dio eu conseguia chamar novamente. Alguem tem alguma ideia?

```
class DioTokenInterceptor extends Interceptor {
  @override
  void onRequest(
    RequestOptions options,
    RequestInterceptorHandler handler,
  ) async {
    handler.next(options);
  }

  @override
  void onResponse(Response response, ResponseInterceptorHandler handler) async {
    super.onRequest(response.requestOptions, RequestInterceptorHandler()),
  }
}
```

**Comment by eduardoflorence** [comment]
@scduarte o `onResponse` não retorna void, ele retorna um Future e o primeiro parâmetro de `super.onResponse` não é a propriedade `requestOptions`. O correto seria assim: 
```dart
@override
Future onResponse(Response response, ResponseInterceptorHandler handler) {
  return super.onResponse(response, handler);
}
```

**Reply by scduarte** [reply]
![image](https://user-images.githubusercontent.com/28739260/194158329-59c2ff05-487c-4fe6-acd9-1d8e8edc1034.png)

**Reply by scduarte** [reply]
As funções retornam void.

**Reply by eduardoflorence** [reply]
Verdade, eu me tomei como base o Readme do package e lá mostra um exemplo com Future.
O que acontece quando você chama super.onResponse? Está dando erro? Essa fluxo não ficaria em loop?
Eu normalmente faço um novo request em resposta a um erro de autenticação, mas nunca chamando o super e sim fazendo uma nova requisição.

**Reply by scduarte** [reply]
Então, no caso, eu queria chamar o onRequest novamente.. mas dentro do onResponse. Pq no onResponse, eu faço o refresh do token, e daí queria chamar o onRequest novamente..pra ser feito a chamada com o token novo.

**Comment by JoaoWitorFelipe** [comment]
Bom, pelo que entendi tais chamando o `onRequest` do pai (super) e não do seu interceptor. Não deveria fazer assim:

```dart
class DioTokenInterceptor extends Interceptor {
  @override
  void onRequest(
    RequestOptions options,
    RequestInterceptorHandler handler,
  ) async {
    handler.next(options);
  }

  @override
  void onResponse(Response response, ResponseInterceptorHandler handler) async {
    // Sem o super
    onRequest(response.requestOptions, RequestInterceptorHandler()),
  }
}
```

Isso faz sentido (não testei o código)?

**Reply by JoaoWitorFelipe** [reply]
Na verdade, faz sentido querer executar novamente a request pelo interceptor :thinking:? isso não me parece correto e nem possível pelo Dio não? Não seria melhor implementar direto na implementação aonde executa os métodos `get` ou `post`? Por exemplo:

(custom-dio.dart)
```dart
response = await dio.get('/test?id=12&name=wendu');
if (response.status == 401) {
   // executa um processo padrão de refresh token
  // chama a mesma função repetindo o processo
}
```

Algo recursivo talvez? 
Seria só uma proposta diferente ao invés de usar o interceptor pra isso, fazer direto na implementação do cliente http.

**Reply by JoaoWitorFelipe** [reply]
Olhando a documentação do Dio, eu encontrei esse plugin https://github.com/rodion-m/dio_smart_retry, que pelo visto é um interceptor que ele usa :sweat_smile:, eu só não sei se ele dá algum callback para poder executar algo mas já tem algo pronto.

**Comment by eduardoflorence** [comment]
@scduarte, dei uma estudada melhor no seu cenário e comparando com o modo que uso aqui para refresh token. Seguem minhas observações:
1) A orientação do author do Dio para um cenário **oauth** (token/refresh token) é utilizar `QueuedInterceptor` ao invés de `Inteceptor`, pois as requisições (requisição inicial/requisição refresh token/repetir requisição inicial) devem ser enfileiradas e não executadas em paralelo.
2) Não é no `onResponse` que você faz o refresh do token e sim no `onError`, verificando se o `statusCode` é 401
3) Você não tem que fazer um `super.onRequest` após o refresh token e sim uma nova requisição usando o método `fetch` do Dio, passando o `requestOptions` que veio no erro.
4) Para a requisição do token é importante fazer uma nova instância do Dio para não entrar em loop por causa do Interceptor.
5) Para ficar mais fácil entender, o trecho de código abaixo foi retirado de um exemplo do próprio repositório do Dio. Nele é usado o `csrfToken` para representar o refresh token, então é só adaptar para o formato que você utiliza:

```dart
void main() async {
  var dio = Dio();
  //  dio instance to request token
  var tokenDio = Dio();
  String? csrfToken;
  dio.options.baseUrl = 'http://www.dtworkroom.com/doris/1/2.0.0/';
  tokenDio.options = dio.options;
  dio.interceptors.add(QueuedInterceptorsWrapper(
    onRequest: (options, handler) {
      print('send request：path:${options.path}，baseURL:${options.baseUrl}');
      if (csrfToken == null) {
        print('no token，request token firstly...');
        tokenDio.get('/token').then((d) {
          options.headers['csrfToken'] = csrfToken = d.data['data']['token'];
          print('request token succeed, value: ' + d.data['data']['token']);
          print(
              'continue to perform request：path:${options.path}，baseURL:${options.path}');
          handler.next(options);
        }).catchError((error, stackTrace) {
          handler.reject(error, true);
        });
      } else {
        options.headers['csrfToken'] = csrfToken;
        return handler.next(options);
      }
    },
    onError: (error, handler) {
      //print(error);
      // Assume 401 stands for token expired
      if (error.response?.statusCode == 401) {
        var options = error.response!.requestOptions;
        // If the token has been updated, repeat directly.
        if (csrfToken != options.headers['csrfToken']) {
          options.headers['csrfToken'] = csrfToken;
          //repeat
          dio.fetch(options).then(
            (r) => handler.resolve(r),
            onError: (e) {
              handler.reject(e);
            },
          );
          return;
        }
        tokenDio.get('/token').then((d) {
          //update csrfToken
          options.headers['csrfToken'] = csrfToken = d.data['data']['token'];
        }).then((e) {
          //repeat
          dio.fetch(options).then(
            (r) => handler.resolve(r),
            onError: (e) {
              handler.reject(e);
            },
          );
        });
        return;
      }
      return handler.next(error);
    },
  ));
}
```

**Reply by renanzdm** [reply]
Também faço assim
